src/compiler/tracing.ts

```diff
@@ -88,38 +92,39 @@ namespace ts.tracing {
         Emit = "emit",
     }
 
-    export function begin(phase: Phase, name: string, args: object) {
-        if (!traceFd) {
-            return;
-        }
-        Debug.assert(fs);
+    export function begin(phase: Phase, name: string, args?: object) {
+        spitEvent("B", phase, name, args);
+    }
 
-        performance.mark("beginTracing");
-        fs.writeSync(traceFd, `{"pid":1,"tid":1,"ph":"B","cat":"${phase}","ts":${1000 * timestamp()},"name":"${name}","args":{ "ts": ${JSON.stringify(args)} }},\n`);
-        performance.mark("endTracing");
-        performance.measure("Tracing", "beginTracing", "endTracing");
+    export function end(phase: Phase, name: string, args?: object) {
+        spitEvent("E", phase, name, args);
     }
 
-    export function end() {
-        if (!traceFd) {
-            return;
-        }
-        Debug.assert(fs);
+    export function instant(phase: Phase, name: string, args?: object) {
+        spitEvent("I", phase, name, args, `"s":"g"`);
+    }
 
-        performance.mark("beginTracing");
-        fs.writeSync(traceFd, `{"pid":1,"tid":1,"ph":"E","ts":${1000 * timestamp()}},\n`);
-        performance.mark("endTracing");
-        performance.measure("Tracing", "beginTracing", "endTracing");
+    // Used for "Complete" (ph:"X") events
+    const completeEvents: { phase: Phase, name: string, args?: object, time: number }[] = [];
+    export function push(phase: Phase, name: string, args?: object) {
+        completeEvents.push({ phase, name, args, time: 1000 * timestamp() });
+    }
+    export function pop() {
+        Debug.assert(completeEvents.length > 0);
+        const { phase, name, args, time } = completeEvents.pop()!;
+        const dur = 1000 * timestamp() - time;
+        spitEvent("X", phase, name, args, `"dur":${dur}`, time);
     }
 
-    export function instant(phase: Phase, name: string, args: object) {
-        if (!traceFd) {
-            return;
-        }
+    function spitEvent(eventType: string, phase: Phase, name: string, args: object | undefined, extras?: string,

```

Comment:
I have to admit that I don't love this name.  Maybe we could just say "writeEvent"?

Reply:
(I *almost* couldn't resist `spewEvent`... :)